### ISSUE TITLE
X-ray models calc() specification

### ISSUE THREAD

**Issue by JohannesBuchner** [sherpa/sherpa#1271]
Hi,

I wrote some custom model classes many years ago by reverse-engineering some sherpa models (such as Powerlaw). Currently I am debugging this old code for other reasons and just want to make sure I did get things right.

I am subclassing ArithmeticModel and overriding calc, startup, teardown and guess. The most important one is calc, for which I have:

def calc(self, params, left, right, *args, **kwargs):

where I believe 

* params are the values of the defined parameters of the model, 
* left is the lower energy for each bin
* right is the upper energy for each bin

The function is supposed to return an array of fluxes, 
**each integrated over the bin from `left` to `right`**.

Is this correct? I am especially unsure about the last part, and could not find it in the documentation.

Cheers,
            Johannes

**Comment by DougBurke** [comment]
TLDR: you are pretty much right although integration depends on the integrate setting and there have been some changes in the class hieararchy

There is some documentation at https://sherpa.readthedocs.io/ - for example https://sherpa.readthedocs.io/en/latest/evaluation/index.html#direct-evaluation-of-the-model and https://sherpa.readthedocs.io/en/latest/model_classes/model.html#model-evaluation but there are complications (because Sherpa is full of complications) such as https://sherpa.readthedocs.io/en/latest/evaluation/integrate.html

https://sherpa.readthedocs.io/en/latest/evaluation/examples.html#model-evaluate-example-pha puts some things together but probably not at the level you need.

So

```python
def calc(self, params, *args, **args):
```

has, as you said

- `params` is the set of parameter values to use for the model (all parameters, not just thawed ones). The order matches the `self.pars` order
- `*args` defines the independent axes to evaluate the model on
- `**kwargs` are currently not used (or rarely used)

In a "normal" Sherpa session (where you are not calling the `calc` method directly) then the `*args` is taken from the `Data` object via `eval_model`/`eval_model_to_fit` - https://sherpa.readthedocs.io/en/latest/data/index.html#evaluating-a-model so it depends on what you expect the model to be used with.

There is now a `ndim` attribute for models that Sherpa uses to ensure we don't mix 1D and 2D models. If it's None then we assume it can be used for anything, otherwise we check when models are combined - e.g.

```
>>> gal + bob
ModelErr: Models do not match: 1D (xsphabs.gal) and 2D (const2d.bob)
```

As an aside: the `ndim` attribute is not used when calling `calc` (ie it is not checked to ensure the number of arguments is correct). It probably should be.

If it's for `DataPHA` data then tracking this down is complicated because responses complicate things - once you've applied a response to a model - which returns a new model - then that new model works in channel units and will internally send the "correct" arguments to the original model. For reference we normally use the https://sherpa.readthedocs.io/en/latest/model_classes/api/sherpa.astro.instrument.RSPModelPHA.html response but there others.

So for DataPHA data `*args*` will be sent as two arrays, the low edge (`lo`) and the upper edge (`hi`) of each bin. It is expected that these bins do not overlap; if they do then strange things could happen when you include XSPEC models in your model expression. We require for bin i

    lo_i < hi_i

but we can have either an "energy" grid, where

    lo_i + 1 >= hi_i

or a wavelength grid which has

   lo_i + 1 <= hi_i

That is, we have a heuristic that energy values are in ascending order and wavelength bins are in descending order [**at least that's my understanding of the behavior and this set up well before astropy was born, never mnd astropy.units].

The `calc` method can do whatever it wants with the `lo` and `hi` arguments, but it is expected that you calculate the model within the lo_i to hi_i range. If you have an "additive" model (using XSPEC terminology) - that is a source of emission - then you are generally going to want to integrate the model over the lo_i to hi_i range. If you have a "multiplicative" model (again, XSPEC terminology) which is one that multiplies the spectrum by a value at a given energy then you can chose lo_i, hi_i, (lo_i + hi_i)/2, or some other scheme and evaluate the model at that value for each bin.

Each model does have an integrate which is meant to tell the model what to do when given (xlo, xhi) rather than just x. The handling of integrate is one part of Sherpa I've been trying to improve, both in documentation and behavior, and it is entirely possible for a model to ignore the `integrate` setting and decide it's always going to integrate over the bin (XSPEC additive models) or just evaluate the model at a single point within each bin (XSPEC multiplicative model)

You can use ArithmeticModel as your base class. There is a new-ish class our models now use, [sherpa.models.model.RegriddableModel1D](https://sherpa.readthedocs.io/en/latest/model_classes/api/sherpa.models.model.RegriddableModel1D.html) but I wouldn't bother worrying about it until you decide you want to use the [regrid support](https://sherpa.readthedocs.io/en/latest/evaluation/regrid.html) as I think there are still gremlins yet to be uncovered there.

If you expect the model to be used for non-PHA data then the above still holds except that  you don't know what the units of the independent axis are (could be km, Angstrom, keV, femto-yards per gallon...).

**Comment by JohannesBuchner** [comment]
Thanks a lot for the detailed response.

Yes, I have seen the .integrate property. Actually this tripped me up more than once, because if you do
`const1d.c * othermodel`
this does not do the same as `xsconstant.c * othermodel`. `const1d.c.integrate=False` does the trick.

**Comment by JohannesBuchner** [comment]
Or allow parameters to be multiplied on?

**Comment by DougBurke** [comment]
I don't understand what you mean?

**Comment by JohannesBuchner** [comment]
`powerlaw.p1 * Parameter('src', 'norm', 1, 0, 10, 0, 10)`:

```
Traceback (most recent call last):
  File "xagnfitter_veryoldmodel.py", line 224, in <module>
    powerlaw.p1 * Parameter('src', 'norm', 1, 0, 10, 0, 10)
  File "/home/user/bin/ciao-4.13/lib/python3.7/site-packages/sherpa-4.13.0-py3.7-linux-x86_64.egg/sherpa/models/model.py", line 587, in func
    return BinaryOpModel(self, rhs, op, opstr)
  File "/home/user/bin/ciao-4.13/lib/python3.7/site-packages/sherpa-4.13.0-py3.7-linux-x86_64.egg/sherpa/models/model.py", line 790, in __init__
    self.rhs = self.wrapobj(rhs)
  File "/home/user/bin/ciao-4.13/lib/python3.7/site-packages/sherpa-4.13.0-py3.7-linux-x86_64.egg/sherpa/models/model.py", line 786, in wrapobj
    return _wrapobj(obj, ArithmeticConstantModel)
  File "/home/user/bin/ciao-4.13/lib/python3.7/site-packages/sherpa-4.13.0-py3.7-linux-x86_64.egg/sherpa/models/model.py", line 1052, in _wrapobj
    return wrapper(obj)
  File "/home/user/bin/ciao-4.13/lib/python3.7/site-packages/sherpa-4.13.0-py3.7-linux-x86_64.egg/sherpa/models/model.py", line 536, in __init__
    val = SherpaFloat(val)
TypeError: float() argument must be a string or a number, not 'Parameter'
```

`Parameter('src', 'norm', 1, 0, 10, 0, 10) * powerlaw.p1`:

```
TypeError: float() argument must be a string or a number, not 'Powerlaw'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "xagnfitter_veryoldmodel.py", line 224, in <module>
    Parameter('src', 'norm', 1, 0, 10, 0, 10) * powerlaw.p1
  File "/home/user/bin/ciao-4.13/lib/python3.7/site-packages/sherpa-4.13.0-py3.7-linux-x86_64.egg/sherpa/models/parameter.py", line 95, in func
    return BinaryOpParameter(self, rhs, op, opstr)
  File "/home/user/bin/ciao-4.13/lib/python3.7/site-packages/sherpa-4.13.0-py3.7-linux-x86_64.egg/sherpa/models/parameter.py", line 549, in __init__
    self.rhs = self.wrapobj(rhs)
  File "/home/user/bin/ciao-4.13/lib/python3.7/site-packages/sherpa-4.13.0-py3.7-linux-x86_64.egg/sherpa/models/parameter.py", line 545, in wrapobj
    return ConstantParameter(obj)
  File "/home/user/bin/ciao-4.13/lib/python3.7/site-packages/sherpa-4.13.0-py3.7-linux-x86_64.egg/sherpa/models/parameter.py", line 519, in __init__
    self.value = SherpaFloat(value)
ValueError: setting an array element with a sequence.
```

BinaryOpModel and BinaryOpParameter could be modified, to wrap with a scale1d thing and return a model.

**Comment by DougBurke** [comment]
Given that we have the Scale1D model I'm not sure it's worth adding support for this particular feature, or am I missing something?

There are some issues with parameter linking (I forget the details) that something like this may be useful for, but I'm not sure.

**Comment by hamogu** [comment]
This is an interesting idea that could simplify building up more complicated models; however, like @DougBurke I somewhat suspect that this is easier addressed through documentation. We can deprecate (or at least no longer advertise) const1d and instead highlight scale1d, which will "just work" in these situations.

I have, after almost 10 years of Sherpa, never had the idea to use a Parameter in that way.
